crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -318,51 +296,66 @@ fn visibility_propagate_system(
 
 fn propagate_recursive(
     parent_visible: bool,
-    visibility_query: &mut Query<(&Visibility, &mut ComputedVisibility, &Parent)>,
-    children_query: &Query<&Children, (With<Parent>, With<Visibility>, With<ComputedVisibility>)>,
+    visibility_query: &mut Query<(&Visibility, &mut InheritedVisibility, &Parent)>,
+    children_query: &Query<&Children, (With<Parent>, With<Visibility>, With<InheritedVisibility>)>,
     entity: Entity,
     expected_parent: Entity,
     // BLOCKED: https://github.com/rust-lang/rust/issues/31436
     // We use a result here to use the `?` operator. Ideally we'd use a try block instead
 ) -> Result<(), ()> {
-    let is_visible = {
-        let (visibility, mut computed_visibility, child_parent) =
-            visibility_query.get_mut(entity).map_err(drop)?;
-        assert_eq!(
+    let (visibility, mut inherited_visibility, child_parent) =
+        visibility_query.get_mut(entity).map_err(drop)?;
+    assert_eq!(
             child_parent.get(), expected_parent,
             "Malformed hierarchy. This probably means that your hierarchy has been improperly maintained, or contains a cycle"
         );
-        let visible_in_hierarchy = (parent_visible && visibility == Visibility::Inherited)
-            || visibility == Visibility::Visible;
-        // reset "view" visibility here ... if this entity should be drawn a future system should set this to true
-        computed_visibility.reset(visible_in_hierarchy);
-        visible_in_hierarchy
-    };
-
-    for child in children_query.get(entity).map_err(drop)?.iter() {
-        let _ = propagate_recursive(is_visible, visibility_query, children_query, *child, entity);
+
+    let is_visible = (parent_visible && visibility == Visibility::Inherited)
+        || visibility == Visibility::Visible;
+    inherited_visibility.set_if_neq(InheritedVisibility(is_visible));
+
+    for &child in children_query.get(entity).map_err(drop)?.iter() {
+        let _ = propagate_recursive(is_visible, visibility_query, children_query, child, entity);
     }
     Ok(())
 }
 
+/// Resets the view visibility of every entity.
+/// Entities that are visible will be marked as such later this frame
+/// by a [`VisibilitySystems::CheckVisibility`] system.
+fn reset_view_visibility(mut query: Query<&mut ViewVisibility>) {
+    for mut view_visibility in &mut query {
```

Comment:
Non-blocking: this may benefit from parallel query iteration: it's a trivial loop with a ton of entities in it.

Reply:
I ran `cargo run --release --example many_foxes`

1. Without par_iter: avg fps 288.0 over 5285 frames.
2. With par_iter: avg fps 281.8 over 5073.

Parallel iteration doesn't seem to have much of an effect, though it's possible that it would scale better for scenes with a huge number of entities.